state/watcher.go

```diff
@@ -893,38 +921,40 @@ func (w *modelFieldChangeWatcher) initial() (set.Strings, error) {
 	defer closer()
 
 	var doc instanceCharmProfileData
-	machineIds := make(set.Strings)
+	watchSet := make(set.Strings)
 	iter := collection.Find(w.members).Iter()
 	for iter.Next(&doc) {
 		// If no members criteria is specified, use the filter
 		// to reject any unsuitable initial elements.
-		if w.members == nil && w.filter != nil && !w.filter(doc.MachineId) {
+		if w.members == nil && w.filter != nil && !w.filter(doc.DocID) {
 			continue
 		}
 
 		if w.completed(doc) {
-			logger.Tracef("field change NOT watching machine %s", doc.MachineId)
+			logger.Tracef("field change NOT watching %s", doc.DocID)
 			continue
 		}
 
-		docField := w.accessor(doc)
-		w.known[doc.MachineId] = docField
-		machineIds.Add(doc.MachineId)
+		docField, returnID := w.accessor(doc)
+		docId := w.backend.localID(doc.DocID)
+		w.known[docId] = docField
+		watchSet.Add(returnID)
 	}
-	if machineIds.Size() > 0 {
-		logger.Debugf("started field change watching of machines %s", machineIds.Values())
+	if watchSet.Size() > 0 {
+		logger.Debugf("started field change watching %s", watchSet.Values())
 	}
-	return machineIds, iter.Close()
+	return watchSet, iter.Close()
 }
 
-func (w *modelFieldChangeWatcher) merge(machineIds set.Strings, change watcher.Change) error {
-	machineId := w.backend.localID(change.Id.(string))
+func (w *modelFieldChangeWatcher) merge(watchSet set.Strings, change watcher.Change) error {
+	docId := w.backend.localID(change.Id.(string))
 	if change.Revno == -1 {
-		if _, ok := w.known[machineId]; ok {
-			logger.Tracef("stopped field change watching for machine %q", machineId)
+		if _, ok := w.known[docId]; ok {
+			logger.Tracef("stopped field change watching for %q", docId)
 		}
-		delete(w.known, machineId)
-		machineIds.Remove(machineId)
+		delete(w.known, docId)
+		// TODO - check this correct.
+		watchSet.Remove(docId)
```

Comment:
So the idea behind this remove, is that we don't look for documents that are no longer needed. If those documents [do come back and they're classified as](https://github.com/juju/juju/pull/9699/files#diff-5a822a48b96dd79a90eda5ba852495c4R977) `!isKnown` (they're not known), then we re-add them in the merge. So in theory this shouldn't be an issue. 

Does that sound reasonable assumption?